airbyte-integrations/connectors/source-salesforce/source_salesforce/streams.py

```diff
@@ -146,20 +156,52 @@ def create_stream_job(self, query: str, url: str) -> Optional[str]:
                 raise error
 
     def wait_for_job(self, url: str) -> str:
-        start_time = pendulum.now()
-        while True:
-            job_info = self._send_http_request("GET", url=url)
-            job_id = job_info.json()["id"]
-            job_status = job_info.json()["state"]
-
-            if job_info.json()["state"] in ["JobComplete", "Aborted", "Failed"]:
+        # using "seconds" argument because self._wait_timeout can be changed by tests
+        expiration_time: DateTime = pendulum.now().add(seconds=int(self._wait_timeout * 60.0))
+        job_status = "InProgress"
+        delay_timeout = 0
+        delay_cnt = 0
+        job_info = None
+        # minimal starting delay is 0.5 seconds.
+        # this value was received empirically
+        time.sleep(0.5)
+        while pendulum.now() < expiration_time:
+            job_info = self._send_http_request("GET", url=url).json()
+            job_status = job_info["state"]
+            if job_status in ["JobComplete", "Aborted", "Failed"]:
                 return job_status
 
-            if pendulum.now() > start_time.add(minutes=self.JOB_WAIT_TIMEOUT_MINS):
-                return job_status
+            if delay_timeout < self.MAX_CHECK_INTERVAL_SECONDS:
+                delay_timeout = 0.5 + math.exp(delay_cnt) / 1000.0
+                delay_cnt += 1
+
+            time.sleep(delay_timeout)
+            job_id = job_info["id"]
+            self.logger.info(
+                f"Sleeping {delay_timeout} seconds while waiting for Job: {self.name}/{job_id}" f" to complete. Current state: {job_status}"
+            )
+
+        self.logger.warning(f"Not wait the {self.name} data for {self._wait_timeout} minutes, data: {job_info}!!")
+        return job_status
 
-            self.logger.info(f"Sleeping {self.CHECK_INTERVAL_SECONDS} seconds while waiting for Job: {job_id} to complete")
-            time.sleep(self.CHECK_INTERVAL_SECONDS)
+    def execute_job(self, query: Mapping[str, Any], url: str) -> str:
+        job_status = "Failed"
+        for i in range(0, self.MAX_RETRY_NUMBER):
+            job_id = self.create_stream_job(query=query, url=url)
+            if not job_id:
+                return None
+            job_full_url = f"{url}/{job_id}"
+            job_status = self.wait_for_job(url=job_full_url)
+            if job_status not in ["UploadComplete", "InProgress"]:
+                break
+            self.logger.error(f"Waiting error. Try to run this job again {i+1}/{self.MAX_RETRY_NUMBER}...")
+            self.abort_job(url=job_full_url)
+            job_status = "Aborted"
+
+        if job_status in ["Aborted", "Failed"]:
+            self.delete_job(url=job_full_url)
+            raise Exception(f"Job for {self.name} stream using BULK API was failed.")
```

Comment:
shouldn't we be retrying this job instead of failing?

Reply:
I was not sure about possible reasons of failures. Unfortunately I didn't succeed to catch any real failed or aborted job. Maybe there are some internal job failure(timeout or resource issues)